libraries/Microsoft.Bot.Builder.AI.QnA/DialogHelper.cs

```diff
@@ -0,0 +1,188 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DialogHelper"/> class.
+        /// Dialog helper to generate dialogs.
+        /// </summary>
+        /// <param name="services">Bot Services.</param>
+        public DialogHelper(QnAMaker services)
+        {
+            QnAMakerActiveLearningDialog = new WaterfallDialog(ActiveLearningDialogName)
+                .AddStep(CallGenerateAnswer)
+                .AddStep(CallTrain)
+                .AddStep(DisplayQnAResult);
+            _services = services;
+        }
+
+        private async Task<DialogTurnResult> CallGenerateAnswer(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            var scoreThreshold = DefaultThreshold;
+            var top = DefaultTopN;
+
+            QnAMakerOptions qnaMakerOptions = null;
+
+            // Getting options
+            if (stepContext.ActiveDialog.State["options"] != null)
+            {
+                qnaMakerOptions = stepContext.ActiveDialog.State["options"] as QnAMakerOptions;
+                scoreThreshold = qnaMakerOptions?.ScoreThreshold != null ? qnaMakerOptions.ScoreThreshold : DefaultThreshold;
+                top = qnaMakerOptions?.Top != null ? qnaMakerOptions.Top : DefaultTopN;
+            }
+            
+            // Storing the context info
+            stepContext.Values[CurrentQuery] = stepContext.Context.Activity.Text;
+            // -Check if previous context is present, if yes then put it with the query
+            // -Check for id if query is present in reverse index
+
+            var response = await _services.GetAnswersAsync(stepContext.Context, qnaMakerOptions).ConfigureAwait(false);
+
+            // TODO: Take this value from GetAnswerResponse 
+            var isActiveLearningEnabled = true;
+            var filteredResponse = response;
+
+            if (isActiveLearningEnabled)
+            {
+                filteredResponse = _services.GetLowScoreVariation(response.ToArray());
+
+                if (filteredResponse.ToList().Count > 1)
+                {
+                    var suggestedQuestions = new List<string>();
+                    foreach (var qna in filteredResponse)
+                    {
+                        suggestedQuestions.Add(qna.Questions[0]);
+                    }
+
+                    // Get hero card activity
+                    var message = CardHelper.GetSuggestionsCard(suggestedQuestions, CardTitle, CardNoMatchText);
+                    await stepContext.Context.SendActivityAsync(message).ConfigureAwait(false);
+
+                    return new DialogTurnResult(DialogTurnStatus.Waiting);
+                }
+            }
+            else
+            {
+                filteredResponse = new QueryResult[] { response.FirstOrDefault() };
+            }
+
+            return await stepContext.NextAsync(new List<QueryResult>(filteredResponse), cancellationToken).ConfigureAwait(false);
```

Comment:
Note that here you are creating another list with the filtered response that I referenced in the comments around multiple copies within lines 84 and 86. This is one additional copy. Lets try to limit copying,.

Reply:
Ping